Productize import app feature

jingcheng16 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

This PR removes the old page for importing an app from another server. The old page was a single page that could only be accessed by manually entering the URL in the browser’s address bar. Users had no way of discovering this page through the UI, and couldn’t know how to download the application source file without reading the wiki.

Before
<img width="1280" alt="image" src="https://github.com/user-attachments/assets/c5001297-4d33-463e-a266-cdda4f4716ec" />

This PR introduces a new menu item under Project Settings, providing a guided UI with clear instructions. Users can now construct the download link directly from the UI and download the application source file without needing to refer to external documentation.

After
<img width="1280" alt="image" src="https://github.com/user-attachments/assets/f827ae0e-d184-4faf-8424-bc1ffc9c1141" />
<img width="1280" alt="image" src="https://github.com/user-attachments/assets/ee808d0d-c32b-4a9b-8e97-1a2fcf62e394" />


## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-17798

- A new view `ImportAppFromAnotherServerView` is created and linked to the left navigation panel. This view renders a simple template with an `<h3>` header and a `<div>` container, which contains a partial view (@biyeun Can I just replace form content in this `<div>` container? Or is there a reason for two views? ) 
- A new partial view `ImportAppStepsView` is created, where the form content will be dynamically replaced. This view is responsible for rendering and processing the step-by-step import flow:
  - Initially, it renders `ConstructAppDownloadLinkForm` (renamed to `ExtractAppInfoForm` in a later commit).
  - Upon submission, the form is POSTed to the `process_first_step` endpoint. If valid, the view switches to rendering the `ImportAppForm`.
  - When `ImportAppForm` is submitted and passes validation, the app import logic is triggered.

Although this summary is written top-down for clarity, the commits are organized bottom-up for easier review:
- It begins with moving a template from the `data_cleaning` app to `hqwebapp` for reuse.
- Then introduces the two forms: `ConstructAppDownloadLinkForm` and `ImportAppForm`.
- Logic that seemed worth reviewing in isolation is split into separate commits.
- Finally, the `ImportAppStepsView` and the entrypoint `ImportAppFromAnotherServerView` are added.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
The core logic for importing an app remains unchanged. The existing import logic is still used when the Step 2 form is submitted and validated.

This flow has been thoroughly tested locally during development and will be tested on staging before deployment.

There is no risk to any existing applications—at worst, if something goes wrong, the newly imported app may be malformed, which the user can delete or fix manually.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

New test added: `corehq.apps.domain.tests.test_forms.TestConstructAppDownloadLinkForm`

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
Move next_action_form from data_cleaning app to hqwebapp so I can reuse it

jingcheng16 committed:
Change encoding type to handle file input

The default is application/x-www-form-urlencoded

jingcheng16 committed:
Create crispy forms for step 1 and step 2

jingcheng16 committed:
Add validation for app_url

jingcheng16 committed:
Construct download url either from passed parameter or from hidden input

jingcheng16 committed:
Add htmx view to handle the replacement of form

jingcheng16 committed:
Add the main view and template that contains the steps view

jingcheng16 committed:
Remove old import app code in app_manager

jingcheng16 committed:
Rename ConstructAppDownloadLinkForm to ExtractAppInfoForm

biyeun commented on corehq/apps/hqwebapp/templates/hqwebapp/crispy/next_action_form.html:
```diff
@@ -7,6 +7,7 @@
     hx-target="#{{ container_id }}"
     hx-disabled-elt="find button"
     hq-hx-action="{{ next_action }}"
+    enctype="multipart/form-data"
```
i think this should be fine, but since this is used by data cleaning, just make sure that those forms aren't affected :)

jingcheng16 replied:
Yes, tested, data cleaning is not affected!

biyeun commented on corehq/apps/domain/templates/domain/partials/import_app_step_1_instruction.html:
```diff
@@ -0,0 +1,18 @@
+
+<div class="alert alert-info">
+  <p>
+    {% trans "This tool helps you import an app from another CommCare server to your current CommCare server." %}
```
nit: i always prefer `{% blocktrans %}` inside of a `<p>` tag, since it's a block element.

jingcheng16 replied:
The text is pretty simple and doesn’t involve any variables, so I went with `trans`, because the [Django docs](https://docs.djangoproject.com/en/dev/topics/i18n/translation/#blocktranslate-template-tag) recommend using `blocktrans` when you need to interpolate variables or translate longer block.
From what I understand, the “block” in `blocktrans` refers to Django’s template block syntax, not an [HTML block element](https://www.w3schools.com/html/html_blocks.asp) — otherwise we’d probably be using `blocktrans` for the `<li>` elements below as well.

jingcheng16 replied:
I re-reviewed the translation tag for the new template and agree there are two sentences that's a bit complicated and might need to use `blocktrans`: 411da11282253d0c7e4072ff17693396ec7b2753

biyeun commented on corehq/apps/domain/forms.py:
```diff
@@ -3122,10 +3124,53 @@ def __init__(self, *args, **kwargs):
+        if not app_url.startswith('https://'):
+            raise forms.ValidationError(_("The URL must start with https://"))
+
+        server_mapping = {
+            'www': 'production',
+            'india': 'india',
+            'eu': 'eu'
+        }
+
+        is_valid_commcare_server = True
+        parsed_url = urllib.parse.urlparse(app_url)
+        netloc = parsed_url.netloc.split(".")
+        subdomain = netloc[0]
+
+        if len(netloc) == 3:
+            domain = netloc[1]
+            suffix = netloc[2]
+            if subdomain not in server_mapping.keys() or domain != 'commcarehq' or suffix != 'org':
+                is_valid_commcare_server = False
+        else:
+            is_valid_commcare_server = False
```
could this be extracted into a staticmethod just to shorten up this whole `clean` function a little bit?

jingcheng16 replied:
e3f358ce5d1da09126b0bc3d73f0ff6493b15cc4

biyeun commented on corehq/apps/domain/forms.py:
```diff
@@ -3122,10 +3124,53 @@ def __init__(self, *args, **kwargs):
+            app_url
+        )
+        if not match:
+            raise forms.ValidationError(_("Invalid app URL format."))
```
same comment above about this being for a specific field

biyeun commented on corehq/apps/domain/views/import_apps.py:
```diff
@@ -0,0 +1,85 @@
+
+    @hq_hx_action('post')
+    def process_second_step(self, request, *args, **kwargs):
+        from corehq.apps.app_manager.views.apps import clear_app_cache
```
was there a circular import here?

jingcheng16 replied:
Yes there is

jingcheng16 commented on corehq/apps/domain/forms.py:
```diff
@@ -3122,10 +3124,53 @@ def __init__(self, *args, **kwargs):
+        app_url = self.cleaned_data['app_url']
+
+        if not app_url.startswith('https://'):
+            raise forms.ValidationError(_("The URL must start with https://"))
```
The error will display for the specific field as long as it is raised in that field's clean method.

jingcheng16 replied:
What do you feel about 898acf87d14f631437dc941c433d3a8541962bf3
I kept the field validation logic in the `clean_app_url()`, but moved the insertion of `cleaned_data` to `clean()`.
I return `app_url` in `clean_app_url()` instead of `None`.  I guess this is the best practice. But I then wipe it in `clean()` method, because I don't want to submit `app_url` to our server. I'm worried about malicious input, user can put anything after the `/` and still pass the validation as long as the beginning part matches our pattern.( Am I overthinking? )

biyeun commented on corehq/tabs/tabclasses.py:
```diff
@@ -2089,6 +2093,17 @@ def sidebar_items(self):
                 section = _get_administration_section(self.domain)
             elif user_can_manage_domain_alerts:
                 section = _get_manage_domain_alerts_section(self.domain)
+            if self.couch_user.can_edit_apps():
+                if section:
+                    section.append({
+                        'title': ImportAppFromAnotherServerView.page_title,
+                        'url': reverse(ImportAppFromAnotherServerView.urlname, args=[self.domain]),
+                    })
+                else:
+                    section = [{
+                        'title': ImportAppFromAnotherServerView.page_title,
+                        'url': reverse(ImportAppFromAnotherServerView.urlname, args=[self.domain]),
+                    }]
```
can we DRY this up by defining `section = []` earlier?

biyeun commented on corehq/apps/app_manager/views/apps.py:
```diff
@@ -619,42 +619,6 @@ def _valid_exchange_record_exists_helper(app_id, records):
     return False
 
 
-@require_can_edit_apps
```
love to see deleted code ^_^

biyeun reviewed (commented):
overall structure looks great! so happy to see another multi-step htmx form!! yayy!

left some comments i'd like to see addressed before giving this a stamp of approval

jingcheng16 committed:
Move next_action_form.html from crispy folder to htmx folder

jingcheng16 committed:
Move next_action_form one level down to forms folder

jingcheng16 committed:
Refactor: add placeholder as a kwarg to crispy.Field

jingcheng16 committed:
Add space between sentences, add period at the end of the sentences

jingcheng16 committed:
Refactor: extract the verify server logic into a static method

jingcheng16 committed:
Refactor: move the cleaned_data code to form level clean()

jingcheng16 commented on corehq/apps/domain/forms.py:
```diff
@@ -3093,3 +3093,89 @@ def __init__(self, request, *args, **kwargs):
+        self.helper.form_tag = False
+        self.helper.layout = crispy.Layout(
+            crispy.HTML(render_to_string('domain/partials/import_app_step_1_instruction.html', {})),
+            crispy.Field('app_url'),
```
71fec4f7ba657e7866e06b47fd1bbd096a678394

jingcheng16 committed:
Add test for non commcare.org url

jingcheng16 commented on corehq/apps/domain/tests/test_forms.py:
```diff
@@ -580,3 +580,132 @@ def test_confidential_forms_autocomplete(self, mock_settings):
+        self.assertEqual(form.cleaned_data['app_id'], '62891a383516c656850cc9c7e7b8d459')
+
+    def test_clean_app_url_with_port_number(self):
+        url = 'https://www.commcarehq.org:443/a/test-domain/apps/view/62891a383516c656850cc9c7e7b8d459/'
```
0df82671e9ad844682d967f010c93356b16469f3

jingcheng16 committed:
reset file pointer before reading app_file

The uploaded file is validated in `clean_app_file()` by calling `json.load(file)`, which advances the file pointer. Later, when attempting to read the file again during app import, `json.load()` fails because the file pointer is already at the end of the stream, and will raise JSONDecodeError

jingcheng16 committed:
use self.render_htmx_redirect from the HqHtmxActionMixin

jingcheng16 committed:
DRY the logic

jingcheng16 commented on corehq/apps/domain/views/import_apps.py:
```diff
@@ -0,0 +1,85 @@
+                return HttpResponse(
+                    status=200,
+                    headers={'HX-Redirect': response.url}
+                )
```
968d6817c7b46c2cd491c3ee3ae44fdb9543735a

jingcheng16 committed:
Remove unused import and run isort

biyeun commented on corehq/apps/domain/forms.py:
```diff
@@ -3224,3 +3242,12 @@ def clean_app_file(self):
+        server_mapping = {
+            'production': 'www',
+            'india': 'india',
+            'eu': 'eu'
```
I still see the mapping itself repeated. can this be extracted to a constant somewhere? reuse the mapping, rather than having the same thing defined in two places

biyeun replied:
ohhh i see...they are not the same, just the same name

```
server_mapping = {
            'www': 'production',
            'india': 'india',
            'eu': 'eu',
        }
```

vs

```
server_mapping = {
            'production': 'www',
            'india': 'india',
            'eu': 'eu',
        }
```

got it...

you can ignore.

jingcheng16 committed:
move logic from clean_app_url to clean

jingcheng16 committed:
use self.add_error and return early to avoid downstream errors

self.add_error() only adds the error — it does not stop execution

jingcheng16 committed:
use blocktrans for complicated sentences

biyeun reviewed (approved):
thanks! this looks good

gherceg reviewed (commented):
Sorry was sitting on a partial review that I should have just submitted. Better late than never?
